iroh-sync/src/keys.rs

```diff
@@ -0,0 +1,283 @@
+    ///
+    /// Will return an error if the input bytes do not represent a valid [`ed25519_dalek`]
+    /// curve point. Will never fail for a byte array returned from [`Self::as_bytes`].
+    /// See [`VerifyingKey::from_bytes`] for details.
+    pub fn from_bytes(bytes: &[u8; 32]) -> anyhow::Result<Self> {
+        Ok(AuthorId(VerifyingKey::from_bytes(bytes)?))
+    }
+}
+
+/// Namespace key of a [`crate::Replica`].
+///
+/// Holders of this key can insert new entries into a [`crate::Replica`].
+/// Internally, a [`Namespace`] is a [`SigningKey`] which is used to sign entries.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Namespace {
+    signing_key: SigningKey,
+}
+
+impl Namespace {
+    /// Create a new [`Namespace`] with a random key.
+    pub fn new<R: CryptoRngCore + ?Sized>(rng: &mut R) -> Self {
+        let signing_key = SigningKey::generate(rng);
+
+        Namespace { signing_key }
+    }
+
+    /// Create a [`Namespace`] from a byte array.
+    pub fn from_bytes(bytes: &[u8; 32]) -> Self {
+        SigningKey::from_bytes(bytes).into()
+    }
+
+    /// Returns the [`Namespace`] byte representation.
+    pub fn to_bytes(&self) -> [u8; 32] {
+        self.signing_key.to_bytes()
+    }
+
+    /// Get the [`NamespaceId`] for this namespace.
+    pub fn id(&self) -> NamespaceId {
+        NamespaceId(self.signing_key.verifying_key())
+    }
+
+    /// Sign a message with this [`Namespace`] key.
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        self.signing_key.sign(msg)
+    }
+
+    /// Strictly verify a signature on a message with this [`Namespace`]'s public key.
+    pub fn verify(&self, msg: &[u8], signature: &Signature) -> Result<(), SignatureError> {
+        self.signing_key.verify_strict(msg, signature)
+    }
+}
+
+/// Identifier for a [`Namespace`]
+///
+/// This is the corresponding [`VerifyingKey`] for a [`Namespace`]. It is used as an identifier, and can
+/// be used to verify [`Signature`]s.
+#[derive(Default, Copy, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct NamespaceId(VerifyingKey);
+
+impl NamespaceId {
+    /// Verify that a signature matches the `msg` bytes and was created with the [`Namespace`]
+    /// that corresponds to this [`NamespaceId`].
+    pub fn verify(&self, msg: &[u8], signature: &Signature) -> Result<(), SignatureError> {
+        self.0.verify_strict(msg, signature)
+    }
+
+    /// Get the byte representation of this [`NamespaceId`].
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        self.0.as_bytes()
+    }
+
+    /// Create from a slice of bytes.
+    ///
+    /// Will return an error if the input bytes do not represent a valid [`ed25519_dalek`]
+    /// curve point. Will never fail for a byte array returned from [`Self::as_bytes`].
+    /// See [`VerifyingKey::from_bytes`] for details.
+    pub fn from_bytes(bytes: &[u8; 32]) -> anyhow::Result<Self> {
+        Ok(NamespaceId(VerifyingKey::from_bytes(bytes)?))
+    }
+}
+
+impl fmt::Display for Author {
```

Comment:
you can avoid these manual impls by using `derive_more::Display` much easier to do

Reply:
https://docs.rs/derive_more/1.0.0-beta.3/derive_more/derive.Display.html

Reply:
OK, can look into this. However I'm not sure if I do like it better to put actual code into the derive attributes, for me that doesn't really look simpler than the impl here.

Reply:
not super important